File: irrexplorer/backends/peeringdb.py
"""
PeeringDB backend integration.

Provides peering and interconnection data from PeeringDB.
"""

import asyncio
import logging
from typing import Any, Dict, List, Optional

import aiohttp

logger = logging.getLogger(__name__)


class PeeringDBClient:
    """Client for PeeringDB API queries."""

    API_BASE_URL = "https://api.peeringdb.com/api"

    def __init__(self, timeout: int = 30):
        """
        Initialize PeeringDB client.

        Args:
            timeout: Request timeout in seconds
        """
        self.timeout = aiohttp.ClientTimeout(total=timeout)

    async def query_asn(self, asn: int) -> Dict[str, Any]:
        """
        Query PeeringDB for ASN information.

        Args:
            asn: Autonomous System Number

        Returns:
            Dictionary with network and peering information
        """
        try:
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                url = f"{self.API_BASE_URL}/net?asn={asn}"
                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        networks = data.get("data", [])
                        if networks:
                            return await self._enrich_network_data(networks[0], session)
                        return {"asn": asn, "error": "Not found"}
                    else:
                        return {
                            "asn": asn,
                            "error": f"API error: {response.status}",
                        }
        except asyncio.TimeoutError:
            logger.error(f"Timeout querying PeeringDB for ASN {asn}")
            return {"asn": asn, "error": "Query timeout"}
        except Exception as e:
            logger.error(f"Error querying PeeringDB for ASN {asn}: {e}")
            return {"asn": asn, "error": str(e)}

    async def query_facility(self, facility_id: int) -> Dict[str, Any]:
        """
        Query PeeringDB for facility information.

        Args:
            facility_id: Facility ID

        Returns:
            Dictionary with facility information
        """
        try:
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                url = f"{self.API_BASE_URL}/fac/{facility_id}"
                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        facilities = data.get("data", [])
                        if facilities:
                            return self._parse_facility(facilities[0])
                        return {"facility_id": facility_id, "error": "Not found"}
                    else:
                        return {
                            "facility_id": facility_id,
                            "error": f"API error: {response.status}",
                        }
        except Exception as e:
            logger.error(f"Error querying PeeringDB for facility {facility_id}: {e}")
            return {"facility_id": facility_id, "error": str(e)}

    async def query_ix(self, ix_id: int) -> Dict[str, Any]:
        """
        Query PeeringDB for Internet Exchange information.

        Args:
            ix_id: Exchange ID

        Returns:
            Dictionary with IX information
        """
        try:
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                url = f"{self.API_BASE_URL}/ix/{ix_id}"
                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        ixs = data.get("data", [])
                        if ixs:
                            return self._parse_ix(ixs[0])
                        return {"ix_id": ix_id, "error": "Not found"}
                    else:
                        return {
                            "ix_id": ix_id,
                            "error": f"API error: {response.status}",
                        }
        except Exception as e:
            logger.error(f"Error querying PeeringDB for IX {ix_id}: {e}")
            return {"ix_id": ix_id, "error": str(e)}

    async def search_networks(self, query: str) -> List[Dict[str, Any]]:
        """
        Search for networks by name or other criteria.

        Args:
            query: Search query string

        Returns:
            List of matching networks
        """
        try:
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                url = f"{self.API_BASE_URL}/net?name__contains={query}"
                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        networks = data.get("data", [])
                        return [self._parse_network_basic(net) for net in networks[:20]]
                    return []
        except Exception as e:
            logger.error(f"Error searching PeeringDB networks: {e}")
            return []

    async def _enrich_network_data(
        self, network: Dict[str, Any], session: aiohttp.ClientSession
    ) -> Dict[str, Any]:
        """Enrich network data with additional PeeringDB information."""
        net_id = network.get("id")

        if not net_id:
            # Return basic network data if no ID
            return self._parse_network_basic(network)

        # Fetch additional data in parallel
        facilities_task = self._fetch_network_facilities(net_id, session)
        ix_connections_task = self._fetch_network_ix_connections(net_id, session)

        facilities, ix_connections = await asyncio.gather(
            facilities_task, ix_connections_task, return_exceptions=True
        )

        # Handle exceptions from gather
        if isinstance(facilities, Exception):
            facilities = []
        if isinstance(ix_connections, Exception):
            ix_connections = []

        return {
            "asn": network.get("asn"),
            "name": network.get("name"),
            "aka": network.get("aka"),
            "website": network.get("website"),
            "looking_glass": network.get("looking_glass"),
            "route_server": network.get("route_server"),
            "irr_as_set": network.get("irr_as_set"),
            "info_type": network.get("info_type"),
            "info_scope": network.get("info_scope"),
            "info_prefixes4": network.get("info_prefixes4"),
            "info_prefixes6": network.get("info_prefixes6"),
            "info_traffic": network.get("info_traffic"),
            "info_ratio": network.get("info_ratio"),
            "info_unicast": network.get("info_unicast"),
            "info_multicast": network.get("info_multicast"),
            "info_ipv6": network.get("info_ipv6"),
            "info_never_via_route_servers": network.get("info_never_via_route_servers"),
            "policy_general": network.get("policy_general"),
            "policy_locations": network.get("policy_locations"),
            "policy_ratio": network.get("policy_ratio"),
            "policy_contracts": network.get("policy_contracts"),
            "facilities": facilities,
            "ix_connections": ix_connections,
            "notes": network.get("notes"),
            "created": network.get("created"),
            "updated": network.get("updated"),
        }

    async def _fetch_network_facilities(
        self, net_id: int, session: aiohttp.ClientSession
    ) -> List[Dict[str, Any]]:
        """Fetch facilities for a network."""
        try:
            url = f"{self.API_BASE_URL}/netfac?net_id={net_id}"
            async with session.get(url) as response:
                if response.status == 200:
                    data = await response.json()
                    return [
                        {
                            "facility_id": fac.get("fac_id"),
                            "name": fac.get("name"),
                            "city": fac.get("city"),
                            "country": fac.get("country"),
                        }
                        for fac in data.get("data", [])
                    ]
        except Exception as e:
            logger.error(f"Error fetching facilities for network {net_id}: {e}")
        return []

    async def _fetch_network_ix_connections(
        self, net_id: int, session: aiohttp.ClientSession
    ) -> List[Dict[str, Any]]:
        """Fetch IX connections for a network."""
        try:
            url = f"{self.API_BASE_URL}/netixlan?net_id={net_id}"
            async with session.get(url) as response:
                if response.status == 200:
                    data = await response.json()
                    return [
                        {
                            "ix_id": ix.get("ix_id"),
                            "name": ix.get("name"),
                            "ipaddr4": ix.get("ipaddr4"),
                            "ipaddr6": ix.get("ipaddr6"),
                            "speed": ix.get("speed"),
                            "is_rs_peer": ix.get("is_rs_peer"),
                            "operational": ix.get("operational"),
                        }
                        for ix in data.get("data", [])
                    ]
        except Exception as e:
            logger.error(f"Error fetching IX connections for network {net_id}: {e}")
        return []

    def _parse_network_basic(self, network: Dict[str, Any]) -> Dict[str, Any]:
        """Parse basic network information."""
        return {
            "asn": network.get("asn"),
            "name": network.get("name"),
            "aka": network.get("aka"),
            "website": network.get("website"),
            "info_type": network.get("info_type"),
            "info_scope": network.get("info_scope"),
        }

    def _parse_facility(self, facility: Dict[str, Any]) -> Dict[str, Any]:
        """Parse facility information."""
        return {
            "id": facility.get("id"),
            "name": facility.get("name"),
            "aka": facility.get("aka"),
            "website": facility.get("website"),
            "clli": facility.get("clli"),
            "rencode": facility.get("rencode"),
            "npanxx": facility.get("npanxx"),
            "address1": facility.get("address1"),
            "address2": facility.get("address2"),
            "city": facility.get("city"),
            "state": facility.get("state"),
            "zipcode": facility.get("zipcode"),
            "country": facility.get("country"),
            "latitude": facility.get("latitude"),
            "longitude": facility.get("longitude"),
            "available_voltage_services": facility.get("available_voltage_services"),
            "diverse_serving_substations": facility.get("diverse_serving_substations"),
            "property": facility.get("property"),
        }

    def _parse_ix(self, ix: Dict[str, Any]) -> Dict[str, Any]:
        """Parse Internet Exchange information."""
        return {
            "id": ix.get("id"),
            "name": ix.get("name"),
            "aka": ix.get("aka"),
            "name_long": ix.get("name_long"),
            "city": ix.get("city"),
            "country": ix.get("country"),
            "region_continent": ix.get("region_continent"),
            "media": ix.get("media"),
            "proto_unicast": ix.get("proto_unicast"),
            "proto_multicast": ix.get("proto_multicast"),
            "proto_ipv6": ix.get("proto_ipv6"),
            "website": ix.get("website"),
            "url_stats": ix.get("url_stats"),
            "tech_email": ix.get("tech_email"),
            "tech_phone": ix.get("tech_phone"),
            "policy_email": ix.get("policy_email"),
            "policy_phone": ix.get("policy_phone"),
        }


# Global client instance
_peeringdb_client: Optional[PeeringDBClient] = None


def get_peeringdb_client() -> PeeringDBClient:
    """Get or create the global PeeringDB client instance."""
    global _peeringdb_client
    if _peeringdb_client is None:
        _peeringdb_client = PeeringDBClient()
    return _peeringdb_client

File: go-backend/internal/datasources/peeringdb.go
package datasources

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

type PeeringDBClient struct {
	httpClient *http.Client
	baseURL    string
}

func NewPeeringDBClient(timeout time.Duration) *PeeringDBClient {
	return &PeeringDBClient{
		httpClient: &http.Client{Timeout: timeout},
		baseURL:    "https://api.peeringdb.com/api",
	}
}

func (c *PeeringDBClient) QueryASN(ctx context.Context, asn int) map[string]any {
	data, err := c.get(ctx, fmt.Sprintf("%s/net?asn=%d", c.baseURL, asn))
	if err != nil {
		return map[string]any{"asn": asn, "error": err.Error()}
	}
	networks := nestedData(data)
	if len(networks) == 0 {
		return map[string]any{"asn": asn, "error": "Not found"}
	}
	network, _ := networks[0].(map[string]any)
	return map[string]any{
		"asn":                          network["asn"],
		"name":                         network["name"],
		"aka":                          network["aka"],
		"website":                      network["website"],
		"looking_glass":                network["looking_glass"],
		"route_server":                 network["route_server"],
		"irr_as_set":                   network["irr_as_set"],
		"info_type":                    network["info_type"],
		"info_scope":                   network["info_scope"],
		"info_prefixes4":               network["info_prefixes4"],
		"info_prefixes6":               network["info_prefixes6"],
		"info_traffic":                 network["info_traffic"],
		"info_ratio":                   network["info_ratio"],
		"info_unicast":                 network["info_unicast"],
		"info_multicast":               network["info_multicast"],
		"info_ipv6":                    network["info_ipv6"],
		"info_never_via_route_servers": network["info_never_via_route_servers"],
		"policy_general":               network["policy_general"],
		"policy_locations":             network["policy_locations"],
		"policy_ratio":                 network["policy_ratio"],
		"policy_contracts":             network["policy_contracts"],
		"notes":                        network["notes"],
		"created":                      network["created"],
		"updated":                      network["updated"],
		"facilities":                   []any{},
		"ix_connections":               []any{},
	}
}

func (c *PeeringDBClient) QueryFacility(ctx context.Context, facilityID int) map[string]any {
	data, err := c.get(ctx, fmt.Sprintf("%s/fac/%d", c.baseURL, facilityID))
	if err != nil {
		return map[string]any{"facility_id": facilityID, "error": err.Error()}
	}
	items := nestedData(data)
	if len(items) == 0 {
		return map[string]any{"facility_id": facilityID, "error": "Not found"}
	}
	facility, _ := items[0].(map[string]any)
	return facility
}

func (c *PeeringDBClient) QueryIX(ctx context.Context, ixID int) map[string]any {
	data, err := c.get(ctx, fmt.Sprintf("%s/ix/%d", c.baseURL, ixID))
	if err != nil {
		return map[string]any{"ix_id": ixID, "error": err.Error()}
	}
	items := nestedData(data)
	if len(items) == 0 {
		return map[string]any{"ix_id": ixID, "error": "Not found"}
	}
	ix, _ := items[0].(map[string]any)
	return ix
}

func (c *PeeringDBClient) SearchNetworks(ctx context.Context, query string) []map[string]any {
	data, err := c.get(ctx, fmt.Sprintf("%s/net?name__contains=%s", c.baseURL, url.QueryEscape(query)))
	if err != nil {
		return []map[string]any{}
	}
	items := nestedData(data)
	results := make([]map[string]any, 0)
	for _, raw := range items {
		item, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		results = append(results, map[string]any{
			"asn":        item["asn"],
			"name":       item["name"],
			"aka":        item["aka"],
			"website":    item["website"],
			"info_type":  item["info_type"],
			"info_scope": item["info_scope"],
		})
		if len(results) >= 20 {
			break
		}
	}
	return results
}

func (c *PeeringDBClient) get(ctx context.Context, endpoint string) (map[string]any, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API error: %d", resp.StatusCode)
	}
	var data map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}
	return data, nil
}

func nestedData(data map[string]any) []any {
	if items, ok := data["data"].([]any); ok {
		return items
	}
	return []any{}
}
